Allow removal of peer certificate to reduce RAM usage

hanno-arm opened the pull request:
__Context:__ This PR belongs to a series of RAM usage improvements targeting X.509, the first two being #2391 and #2411. The PR is based on #2391 (which is based on #2411).

__Goal:__ Save RAM by not permanently storing the peer's CRT chain unless needed by the user.

__Summary:__ The PR adds a new compile-time option `MBEDTLS_SSL_KEEP_PEER_CERTIFICATE`, enabled by default, controlling the availability of the existing public API `mbedtls_ssl_get_peer_cert()` which allows the user to inspect the peer's CRT at any time during the connection. If it is disabled, the PR modifies the stack to parse, copy and keep the peer's CRT only to the extended needed to preserve functionality and security (see below).

__Background:__ After parsing, the peer's CRT is mainly used in the following places:
1. Certificate verification
1. The underlying public key is always needed at some point to directly or indirectly verify that the peer possesses the corresponding private key. For example, there are signatures in `ServerKeyExchange` and `CertificateVerify` to be verified, or a PMS to be generated in RSA-based key exchanges.
1. Mitigation of the [triple handshake attack](https://mitls.org/pages/attacks/3SHAKE): The peer's CRT must not change during renegotiation, and to that end, the peer's CRT is currently stored as part of `mbedtls_ssl_session`. 
1. The user can query it via `mbedtls_ssl_get_peer_cert()`.

__Approach:__ Remove the peer CRT from the session structure while adapting the above uses as follows:
1. Verification doesn't need the CRT to be part of the long-lived session structure, as it's local to the handling of the incoming `Certificate` state. We can parse and verify the peer's chain in place when reading it from the input buffer, and free all involved structures afterwards.
1. Store a handshake-local copy of the peer's public key which is extracted from the peer's CRT after the parsing and verification step, and free it as soon as it's no longer needed. For example, during an (EC)DHE exchange, the server's `ServerKeyExchange` signature is verified before performing the RAM-expensive (EC)DHE computation, so the peak RAM usage drops if we free the server's public key after handling the `ServerKeyExchange`.
1. Instead of storing the entire CRT in the session structure, store a digest only.

__Technical steps:__ The PR proceeds along the following steps, with the removal of the peer CRT field happening at the very end only:
1. Take the opportunity to restructure and simplify the handling of the incoming certificate state. Several helper functions are introduced for coordination (do we expect a CRT?), no-CRT notification from the client, and verification. 
1. Add fields to `mbedtls_ssl_session` for storing the digest of the peer's CRT. Compute and store the digest after CRT parsing and verification, and mitigate the triple handshake attack by comparing hashes, not the actual CRTs. This leaves the existing peer CRT field in `mbedtls_ssl_session` untouched.
1. Session cache and session tickets need to be adapted to any change of `mbedtls_ssl_session`. Adapt them to store the new digest field in place of the peer CRT field if `MBEDTLS_SSL_KEEP_PEER_CERTFICATE` is disabled. The peer CRT field continues to remain in `mbedtls_ssl_session`, but it's not kept during session resumption. Note that the only use of the peer CRT across session resumption apart from the user calling `mbedtls_ssl_get_peer_cert()` is to mitigate the triple handshake attack, so while this change removes the peer CRT from ticket and cache, the library's functionality is maintained.
1. Add a PK context to `mbedtls_ssl_handshake_params` to hold the peer's public key. Extract it after parsing and verifying the CRT. Go through all functions using the peer's public key, and use the new PK context instead of the PK context embedded in the peer's CRT (which is still there).
1. Actually remove the peer CRT from `mbedtls_ssl_session`, and the final adaptations, like modifying `mbedtls_ssl_session_{init/free}()`, or disabling some tests which grep for CRT info output from `ssl_client2` and `ssl_server2`.

__Todo:__ 
- [x] Add tests to `all.sh`
- [x] Enable `MBEDTLS_SSL_KEEP_PEER_CERTIFICATE` by default. It is currently disabled to have the CI test it.
- [x] Free peer's public key as soon as it's no longer needed. So far, it's freed at the end of the handshake only.

hanno-arm commented:
Travis failed because of wrong compile-time guards around a helper function.

hanno-arm commented:
CI succeeded on default config with `MBEDTLS_SSL_KEEP_PEER_CERTIFICATE` disabled.

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -5613,9 +5618,7 @@ static int ssl_parse_certificate_chain( mbedtls_ssl_context *ssl )
-                mbedtls_x509_crt_free( ssl->session_negotiate->peer_cert );
-                mbedtls_free( ssl->session_negotiate->peer_cert );
-                ssl->session_negotiate->peer_cert = NULL;
+                ssl_clear_peer_cert( ssl->session );
```
Thank your for spotting! As far as I see, the previous version, introduced in https://github.com/ARMmbed/mbedtls/pull/2421/commits/825a78e4101cb62ceb593d8db122a16c421b3201, was wrong.

mpg replied:
Good point. After looking at this in context, I agree that the present version is correct.

hanno-arm commented:
Jenkins failed because of timing on FreeBSD only.

k-stachowiak commented on tests/ssl-opt.sh:
```diff
@@ -2830,6 +2830,7 @@ run_test    "Authentication: send CA list in CertificateRequest, client self sig
 
 # Tests for certificate selection based on SHA verson
 
+requires_config_enabled MBEDTLS_SSL_KEEP_PEER_CERTIFICATE
```
Many tests get disabled here for the "non-keep" builds, whereas some unit tests (in test suites) have been added, but in smaller quantity. How does this impact the test coverage in the "non-keep" case? Is it equally tested, or, if it is difficult in the current setup (the way the client and server are implemented), maybe we should at least plan adding more tests, potentially involving client and server modifications?

hanno-arm replied:
@k-stachowiak I've re-introduced all the tests that were previously disabled, by getting the CRT info the tests rely on from a verify callback instead of a call to `mbedtls_ssl_get_peer_cert()` at the end of the handshake.

hanno-arm commented on ChangeLog:
```diff
@@ -2,6 +2,23 @@ mbed TLS ChangeLog (Sorted per branch, date)
 
+Features
+   * Add a new X.509 API call `mbedtls_x509_parse_der_nocopy()`
+     which allows copy-less parsing of DER encoded X.509 CRTs,
```
Yes, that's better. I'll use "in-place".

hanno-arm replied:
Oh, actually this is part of a the separate PR #2411 which is already approved and running through the CI, so I'd prefer to keep things as they are.

hanno-arm commented on include/mbedtls/config.h:
```diff
@@ -1330,6 +1330,28 @@
+/**
+ * \def MBEDTLS_SSL_KEEP_PEER_CERTIFICATE
+ *
+ * This option controls the presence of the API mbedtls_ssl_get_peer_cert()
```
What about "availability"?

k-stachowiak replied:
Even better :)

hanno-arm replied:
Done.

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -5525,124 +5554,392 @@ static int ssl_parse_certificate_chain( mbedtls_ssl_context *ssl )
-            mbedtls_ssl_send_alert_message( ssl, MBEDTLS_SSL_ALERT_LEVEL_FATAL,
-                                            MBEDTLS_SSL_ALERT_MSG_ACCESS_DENIED );
-            return( MBEDTLS_ERR_SSL_BAD_HS_CERTIFICATE );
+            /* NOTE: Is it intentional that we set verify_result
```
Good point - I think in the meantime I found a mail thread on this topic; it's not intentional, but we cannot simply change it because it's an API break. I'll remove the comment because we're aware of it.

hanno-arm replied:
Done.

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -5723,154 +6022,71 @@ int mbedtls_ssl_parse_certificate( mbedtls_ssl_context *ssl )
+    /* Pass ownership to session structure. */
+    ssl->session_negotiate->peer_cert = chain;
+    chain = NULL;
+#endif /* MBEDTLS_SSL_KEEP_PEER_CERTIFICATE */
```
Ok, I'll fix it.

hanno-arm replied:
No, actually it's intended; the comment at the `#endif` indicates the condition on the block that's being closed, which is `MBEDTLS_SSL_KEEP_PEER_CERTIFICATE`.

hanno-arm commented on include/mbedtls/ssl_internal.h:
```diff
@@ -311,8 +311,15 @@ struct mbedtls_ssl_handshake_params
         ssl_ecrs_crt_vrfy_sign,         /*!< CertificateVerify: pk_sign()   */
     } ecrs_state;                       /*!< current (or last) operation    */
+#if defined(MBEDTLS_X509_CRT_PARSE_C)
+    mbedtls_x509_crt *ecrs_peer_cert;        /*!< The peer's CRT chain.     */
```
Will fix that.

hanno-arm replied:
Done.

hanno-arm commented on include/mbedtls/ssl_internal.h:
```diff
@@ -311,8 +311,15 @@ struct mbedtls_ssl_handshake_params
         ssl_ecrs_crt_vrfy_sign,         /*!< CertificateVerify: pk_sign()   */
     } ecrs_state;                       /*!< current (or last) operation    */
+#if defined(MBEDTLS_X509_CRT_PARSE_C)
+    mbedtls_x509_crt *ecrs_peer_cert;        /*!< The peer's CRT chain.     */
```
That looks like a bug - in a PSK-only configuration with restartable ECC enabled (which doesn't make any sense, but it's possible), the library would fail to build. I will fix it - thanks for spotting!

hanno-arm replied:
Actually, it's not a bug, but rather by luck: `MBEDTLS_SSL__ECP_RESTARTABLE` is set only if `MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED` is enabled, which requires `MBEDTLS_X509_CRT_PARSE_C` to be set.

hanno-arm replied:
I think it makes sense to remove the `MBEDTLS_X509_CRT_PARSE_C` guard here, then.

hanno-arm replied:
Removed.

hanno-arm commented on library/ssl_cli.c:
```diff
@@ -2321,6 +2335,13 @@ static int ssl_get_ecdh_params_from_cert( mbedtls_ssl_context *ssl )
+    /* We don't need the peer's public key anymore. Free it,
+     * so that more RAM is available for upcoming expensive
+     * operations like ECDHE. */
+    mbedtls_pk_free( peer_pk );
```
The point here is that `peer_pk` is embedded in `mbedtls_ssl_handshake_params` as `mbedtls_pk_context`, not a pointer. Calling `mbedtls_xxx_free()` multiple times on a context structure is always fine.

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -295,6 +297,21 @@ static int ssl_session_copy( mbedtls_ssl_session *dst, const mbedtls_ssl_session
+#else /* MBEDTLS_SSL_KEEP_PEER_CERTIFICATE */
+    if( src->peer_cert_digest != NULL )
+    {
+        dst->peer_cert_digest_len = src->peer_cert_digest_len;
```
Not too pedantic at all, it's good hygiene to do it in the way you suggest.

hanno-arm replied:
Done.

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -5435,63 +5459,68 @@ int mbedtls_ssl_write_certificate( mbedtls_ssl_context *ssl )
+#if defined(MBEDTLS_SSL_RENEGOTIATION) && defined(MBEDTLS_SSL_CLI_C)
+
+#if defined(MBEDTLS_SSL_KEEP_PEER_CERTIFICATE)
+static int ssl_check_peer_crt_unchanged( mbedtls_ssl_context *ssl,
```
I would prefer to keep the present version, unless you insist.

k-stachowiak replied:
No, I don't, it is fine :)

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -2872,18 +2896,38 @@ int mbedtls_ssl_get_max_out_record_payload( const mbedtls_ssl_context *ssl );
- * \return         the current peer certificate
+ * \brief          Return the peer certificate from the current connection.
+ *
+ *                 For ciphersuites not using certificate-based peer
```
Done.

hanno-arm commented:
@k-stachowiak Thank you very much for your review; I have addressed your comments. Could you please re-review?

hanno-arm commented:
@k-stachowiak I believe to have addressed your comment on `ssl-opt.sh`. Please re-review.

mpg commented:
The pre-requisite PRs #2391 and #2411 have been merged, so removing "need: preceding PR".

mpg commented:
(Changed base twice in order to refresh github's cache.)

mpg commented:
@hanno-arm `tests/scripts/check-generated-files.sh` failed in Travis. Please investigate and fix.

hanno-arm commented:
@mpg Yes, will do - I don't expect it to be significant, to if you don't mind, please don't hold your review for that.

mpg commented:
Sure, most probably a simple matter of re-running the appropriate script. Don't worry, I'm not letting that delay my review :)

mpg commented on include/mbedtls/x509_crt.h:
```diff
@@ -70,6 +70,7 @@ typedef struct mbedtls_x509_crt
     mbedtls_x509_time valid_from;       /**< Start time of certificate validity. */
     mbedtls_x509_time valid_to;         /**< End time of certificate validity. */
 
+    mbedtls_x509_buf pk_raw;
```
Note: this changes the ABI. Not a problem, but needs to be tracked.

hanno-arm replied:
Actually, we only need this for `!MBEDTLS_SSL_KEEP_PEER_CERTIFICATE`, so we could guard this change accordingly to not change the ABI in the default configuration. WDYT?

mpg replied:
Don't we already have an ABI in libmbedx509 change from one of the preceding PRs? If we do, then I'm inclined to no care about this one. Otherwise, yes, let's use guards so that it only happens in non-default configs.

Btw, the guards might still be useful to save some RAM in the default config, though.

hanno-arm replied:
Yes, we already have an ABI change in another X.509 PR, so considering that, a further change does no harm. However, I think it's still cleaner to not add this field if we don't need it, so I'll remove it.

hanno-arm replied:
@mpg Sorry, I have to withdraw that; the X.509 module shouldn't depend on a SSL configuration. I'll leave the field unconditional.

mpg replied:
Excellent point, thanks for noticing.

mpg commented on library/ssl_cli.c:
```diff
@@ -2183,8 +2183,8 @@ static int ssl_write_encrypted_pms( mbedtls_ssl_context *ssl,
-        MBEDTLS_SSL_DEBUG_MSG( 2, ( "certificate required" ) );
-        return( MBEDTLS_ERR_SSL_UNEXPECTED_MESSAGE );
+        /* Should never happen */
+        return( MBEDTLS_ERR_SSL_INTERNAL_ERROR );
```
Minor: returning internal error is usually preceded by `MBEDTLS_SSL_DEBUG_MSG( 1, ( "should never happen" ) );`: if it ever happens, we want to easily know on which line (with `DEBUG_C` enabled).

hanno-arm replied:
Ok, I'll add it in.

hanno-arm replied:
Done.

mpg commented on library/ssl_srv.c:
```diff
@@ -4003,13 +4003,28 @@ static int ssl_parse_certificate_verify( mbedtls_ssl_context *ssl )
+        ssl->state++;
+        return( 0 );
+    }
+#else /* MBEDTLS_SSL_KEEP_PEER_CERTIFICATE */
```
Minor: I think it would be clearer to isolate the decision-making (are we going to skip?) from the execution (inc state and return early), as currently having (visually) 3 `state++` in the function hinders readability. Could be done later (as part of MPS prepartion with a `coordinate` function?)

hanno-arm replied:
Yes, I fully agree, and #1634 does that, see https://github.com/ARMmbed/mbedtls/blob/62594db239cf337ed5498f1b9e11796dfd679a76/library/ssl_srv.c#L4259. Given the time pressure, I'd prefer to not change anything in this PR. Are you ok with that?

mpg replied:
Sure! As I said, most of my remarks don't have to result in code changes now - except the two things marked "major" (and even perhaps only one of them, if you can convince me about the one in `ssl_client2.c`).

mpg commented on library/ssl_tls.c:
```diff
@@ -5621,7 +5621,13 @@ static int ssl_parse_certificate_chain( mbedtls_ssl_context *ssl,
+#if defined(MBEDTLS_SSL_KEEP_PEER_CERTIFICATE)
         ret = mbedtls_x509_crt_parse_der( chain, ssl->in_msg + i, n );
+#else
+        /* If we don't need to store the CRT chani permanently, parse
```
Typo: chani

hanno-arm replied:
Thanks, I will fix it.

hanno-arm replied:
Done.

mpg commented on programs/ssl/ssl_client2.c:
```diff
@@ -438,6 +438,8 @@ static int my_send( void *ctx, const unsigned char *buf, size_t len )
 }
 
 #if defined(MBEDTLS_X509_CRT_PARSE_C)
+static unsigned char peer_crt_info[1024] = { 0 };
```
Major: It seems to me that this buffer needs to be zeroized not only here, but every time the client is preparing for a new handshake (for example when the reconnect option is used), however I can't see any code doing that later in the file.

hanno-arm replied:
In an ordinary run, I don't think it would need zeroization at all because it's always copied from `buf` in `my_verify()` - isn't it? Yet I found is safer to initialize it in any case to have it represent a well-formed null-terminated string at any time.

mpg replied:
What if a the first handshake uses a certificate but the second (in case of reconnect) doesn't? In that case `my_verify()` won't be called in the second handshake, so after the second handshake `peer_crt_info` will contain the cert info from the first handshake, while it should arguably be the empty string at this point. Am I missing something?

hanno-arm replied:
@mpg Yes, good point! I thought about whether or not `peer_crt_info` would always be a well-formed string, but not whether it could become outdated. I'll change it.

hanno-arm replied:
Done.

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -5655,11 +5655,49 @@ static int ssl_srv_check_client_no_crt_notification( mbedtls_ssl_context *ssl )
+ */
+#define SSL_CERTIFICATE_EXPECTED 0
+#define SSL_CERTIFICATE_SKIP     1
+static int ssl_parse_certificate_coordinate( mbedtls_ssl_context *ssl,
```
It's a middle ground - I didn't take everything from MPS I think, but I think it was worth it to make some structural improvements along the way here.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -783,6 +783,22 @@ typedef int mbedtls_ssl_async_resume_t( mbedtls_ssl_context *ssl,
 typedef void mbedtls_ssl_async_cancel_t( mbedtls_ssl_context *ssl );
 #endif /* MBEDTLS_SSL_ASYNC_PRIVATE */
 
+#if defined(MBEDTLS_X509_CRT_PARSE_C)
```
Ok, I'll add `MBEDTLS_SSL_KEEP_PEER_CERTIFICATE` to indicate the scope of these `#define`s. I'll also used  `MBEDTLS_KEY_EXCHANGE__WITH_CERT__ENABLED` instead of `MBEDTLS_X509_CRT_PARSE_C`.

hanno-arm replied:
Done.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -783,6 +783,22 @@ typedef int mbedtls_ssl_async_resume_t( mbedtls_ssl_context *ssl,
+#define MBEDTLS_SSL_PEER_CERT_DIGEST_DFL_TYPE MBEDTLS_MD_SHA1
+#define MBEDTLS_SSL_PEER_CERT_DIGEST_DFL_LEN  20
+#else
+#error "Bad configuration - need SHA-1, SHA-256 or SHA-512 enabled to compute digest of peer CRT."
```
Ok, I'll move it.

hanno-arm replied:
Done.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -798,7 +814,14 @@ struct mbedtls_ssl_session
+     *  changes during renegotiation, mitigating the triple handshake attack. */
+    unsigned char *peer_cert_digest;
+    size_t peer_cert_digest_len;
+    mbedtls_md_type_t peer_cert_digest_type;
```
We need to store this explicitly because the hash might change across session resumption. The digest length is redundant, but I felt it was worth the 4 Bytes of RAM compared to the code to extract it.

mpg replied:
Ok, thanks for the clarification, I did not notice the point about session resumption. I'm happy with the code as it stands then.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -798,7 +814,14 @@ struct mbedtls_ssl_session
+#if !defined(MBEDTLS_SSL_KEEP_PEER_CERTIFICATE)
+    /*! The digest of the peer's end-CRT. This must be kept to detect CRT
+     *  changes during renegotiation, mitigating the triple handshake attack. */
+    unsigned char *peer_cert_digest;
```
@mpg I agree with you, it would be better to embed the digest here; that would also make session serialization simpler. That said, however, that might lead to larger changes in the session ticket module - I'll have a look.

mpg replied:
Yes, actually I think it's probably safer not to make such changes at this point, but I just wanted to point that out for the future.

hanno-arm replied:
Modifying `mbedtls_ssl_session` is an API break - can we change the API in a non-default configuration? If so, I'd probably prefer to leave the change for later; if not, let's do it now.

mpg replied:
I think it's a grey area. I'd rather leave the change for later (we already have an item to optimize structures in our internal pending API changes list, feel free to expand it with info about this) than having a discussion about what's an acceptable API change under time pressure.

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -270,8 +270,8 @@ static unsigned int ssl_mfl_code_to_length( int mfl )
 
-#if defined(MBEDTLS_SSL_CLI_C)
-static int ssl_session_copy( mbedtls_ssl_session *dst, const mbedtls_ssl_session *src )
+int mbedtls_ssl_session_copy( mbedtls_ssl_session *dst,
```
The guard is removed because `mbedtls_ssl_session_copy()` is now used by the server-side session cache.

mpg replied:
Oh, right. Thanks for clarifying that.

mpg commented on library/ssl_ticket.c:
```diff
@@ -158,9 +158,11 @@ int mbedtls_ssl_ticket_setup( mbedtls_ssl_ticket_context *ctx,
+ *  0       .   n-1     session structure, n = sizeof(mbedtls_ssl_session)
+ *  n       .   n+2     peer_cert length = m (0 if no certificate)
+ *  n+3     .   n+2+m   peer cert ASN.1
+ *  n+3+m   .   n+3+m   length of peer certificate digest = k (0 if n digest)
```
Note: this is still relevant even if github marks is as outdated (the typo was only moved to another line).

hanno-arm replied:
Fixed.

hanno-arm commented on library/ssl_ticket.c:
```diff
@@ -194,11 +199,32 @@ static int ssl_save_session( const mbedtls_ssl_session *session,
     *p++ = (unsigned char)( ( cert_len >> 16 ) & 0xFF );
     *p++ = (unsigned char)( ( cert_len >>  8 ) & 0xFF );
     *p++ = (unsigned char)( ( cert_len       ) & 0xFF );
+    left -= 3;
```
Maybe I misunderstand you, but in the latest version that's the case I think.

mpg replied:
It is. I should have deleted this comment when reviewing later commits, but I forgot.

hanno-arm commented:
@mpg Thank you for your review, I believe to have addressed your comments. Could you please re-review?

@k-stachowiak Could you please review the latest commits, too?

mpg commented:
Marking this as approved for design as I found no design flaw in my review.

mpg commented on programs/ssl/ssl_client2.c:
```diff
@@ -2044,6 +2045,8 @@ int main( int argc, char *argv[] )
 
         mbedtls_printf( "  . Reconnecting with saved session..." );
 
+        memset( peer_crt_info, 0, sizeof( peer_crt_info ) );
```
Thanks for adding this. However, there is another code path that reconnects (7b in the comment) which I think need this as well. (Found by looking for other calls to `mbedtls_ssl_session_reset()`.)

hanno-arm replied:
Sorry for that! Fixed.

mpg commented on library/ssl_cli.c:
```diff
@@ -2188,6 +2188,7 @@ static int ssl_write_encrypted_pms( mbedtls_ssl_context *ssl,
     if( ssl->session_negotiate->peer_cert == NULL )
     {
         /* Should never happen */
+        MBEDTLS_SSL_DEBUG_MSG( 1, ( "should never happen" ) );
```
Sorry, should have mentioned in my first comment: there are other occurrences of the same just below.

hanno-arm replied:
Fixed.

hanno-arm commented:
@mpg Thank for you for the re-review, I've addressed your comments.

@mpg @k-stachowiak Please re-review.

hanno-arm commented:
@k-stachowiak @mpg I had to rebase this PR on top of `development` to fix the CI failure. Could you please re-review? The rebase is entirely conflict-free, and I don't see any places where we'd need adaptation to the PSA changes.

mpg commented:
@hanno-arm There are CI failures that look serious and need fixing.

1. `tests/scripts/test-ref-configs.pl config-ccm-psk-tls1_2.h` fails with:
```
/var/lib/build/programs/ssl/ssl_client2.c:2150:43: error: use of undeclared identifier 'peer_crt_info'

        memset( peer_crt_info, 0, sizeof( peer_crt_info ) );

                                          ^

/var/lib/build/programs/ssl/ssl_client2.c:2150:17: error: use of undeclared identifier 'peer_crt_info'

        memset( peer_crt_info, 0, sizeof( peer_crt_info ) );

                ^

/var/lib/build/programs/ssl/ssl_client2.c:2223:43: error: use of undeclared identifier 'peer_crt_info'

        memset( peer_crt_info, 0, sizeof( peer_crt_info ) );

                                          ^

/var/lib/build/programs/ssl/ssl_client2.c:2223:17: error: use of undeclared identifier 'peer_crt_info'

        memset( peer_crt_info, 0, sizeof( peer_crt_info ) );

                ^

4 errors generated.
```
(Found by Jenkins  cmake-asan-clang-debian-9-i386, cmake-asan-clang-debian-9-x64, cmake-full-gcc-debian-9-i386 and Travis Clang.)

2. A variant was found by the Travis GCC job and Jenkins cmake-full-gcc-debian-9-x64 with `tests/scripts/test-ref-configs.pl` now failing in `config-thread.h` (probably just because Perl randomize hashes enumeration so a different failure was encountered first) with:
```
/home/travis/build/ARMmbed/mbedtls/programs/ssl/ssl_client2.c: In function ‘main’:
/home/travis/build/ARMmbed/mbedtls/programs/ssl/ssl_client2.c:2150:17: error: ‘peer_crt_info’ undeclared (first use in this function)
         memset( peer_crt_info, 0, sizeof( peer_crt_info ) );
                 ^
/home/travis/build/ARMmbed/mbedtls/programs/ssl/ssl_client2.c:2150:17: note: each undeclared identifier is reported only once for each function it appears in
```

mpg commented:
@hanno-arm Looks like the fix is simple enough: this identifier is only declared when `MBEDTLS_X509_CRT_PARSE_C` is defined, so the calls to zeroize should be guarded the same way. Sorry for not catching that during review.

hanno-arm commented:
@mpg Thanks for spotting, and sorry for not noticing this earlier.

@mpg @k-stachowiak Could you please re-review the latest commit?

mpg commented:
This is a new feature we want, has two approvals and a passing CI, a ChangeLog entry and new tests in `all.sh`, so it's ready for merge.

hanno-arm commented:
Release job failed because of unused variable warning in client-only build.

mpg commented:
The last commit has been approved and passed the CI. This is now ready for merge again.
